### PR TITLE
Add `carefully`, `errormessage`, allow early returns from `report` and `every`

### PR DESCRIPTION
**Pull request by mrerrormessage**
See also https://github.com/NetLogo/NetLogo-Headless/pull/30 and https://github.com/NetLogo/Tortoise/pull/109. This was opened because the PR for 109 points to the wrong branch, and you can't edit base branches after opening a PR.

[Review on Reviewable](https://reviewable.io/reviews/netlogo/tortoise/125)

### CODE REVIEW

**Comment by TheBizzle** [comment]
I attempted to get this working with the latest Headless artifact, but tests are failing.  I attempted to bisect, but all of the old commits are pointing to a version of Headless that was never published to BinTray.

You can trigger the failing tests by running this in SBT: `testOnly *Commands -- -z Breeds::SetBreedOfUnbreededTurtleWithVar -z Patch::PatchBoundariesAndFloatingPointMagic`.

**Comment by mrerrormessage** [comment]
Yes, I was waiting on Headless to update the version in here. I'll do that this afternoon or monday.

**Comment by TheBizzle** [comment]
If it's not clear, the commit that I pushed to this branch already includes the latest Headless `.jar`.  The tests fail, nonetheless.

**Comment by TheBizzle** [comment]
Ping.

**Comment by mrerrormessage** [comment]
I had forgotten about this PR while you were away. I'll revisit and get a revision back for review.

**Comment by mrerrormessage** [comment]
Updated by merging in master. I know rebasing is preferred to merging, but that was difficult for the same reason bisect was, so I elected to merge.

**Comment by mrerrormessage** [comment]
Tests are passing, but the dependency checker is failing the build for `Shell.scala`, which is unchanged between this branch and master. What do you want me to do?

**Comment by TheBizzle** [comment]
> the dependency checker is failing the build for `Shell.scala`, which is unchanged between this branch and master

The issue is because `Depend.scala` was changed.  You can put the Tortoise Shell in new package that's allowed to depend on `workspace`.

**Comment by mrerrormessage** [comment]
Updated, along with changes to headless, to make the dependency checker passing. I had to break out a new package for the shared test code as well to break Tortoise's dependency on HeadlessWorkspace (well, except for DockingFixture, but that isn't checked). See https://github.com/NetLogo/NetLogo-Headless/pull/33. Note that the test is failing because I haven't yet published the NetLogo-Headless artifact. Locally, all tests are passing.

File: resources/test/dumps/Sandpile.js [hunk 1]
```diff
@@ -245,9 +245,11 @@ function dropPatch() {
     return ListPrims.oneOf(world.patches());
   }
   if ((Prims.equality(world.observer.getGlobal('drop-location'), "mouse-click") && MousePrims.isDown())) {
-    Prims.every(0.3, function() {
+    var dropPatch_0;
+    if (Prims.everyTimerElapsed(0.3, dropPatch_0)) {
       return world.getPatchAt(MousePrims.getX(), MousePrims.getY());
-    }, 'dropPatch-0');
+      dropPatch_0 = new Prims.everyTimer();
+    }
```
Comment (TheBizzle) [inline resources/test/dumps/Sandpile.js]: This implementation of `every` is incorrect—twice!  The code for setting the `every` timer here is occurring on a line that will never be executed (because of the `return` on the preceding line).  Even if this line were moved up, though, this would still be incorrect, because `dropPatch_0` will not persist past the end of the block in which the `everyTimerElapsed` occurs.  The way that this primitive is implemented, I don't see any way for the timer argument to `everyTimerElapsed` to ever be anything but `undefined`, even if `dropPatch_0 = new Prims.everyTimer();` *were* executed.

Simply put, `every` cannot work if its timer is scoped to the same block as the `everyTimerElapsed` call.

Comment (mrerrormessage) [inline resources/test/dumps/Sandpile.js]: It does pass the tests which I wrote for every, but they apparently don't cover all possible behavior. This implementation of every does work in loops (at least loops that don't have early exit), which is (unsurprisingly) what the tests of every are based around.

In NetLogo-Headless, `every` is scoped by command, agent, and job. It is scoped by command here (although that scoping has some problems). I can fix the problems and add in agent-scoping. What is the tortoise analog of a job that I should use for scoping every?

**Comment by mrerrormessage** [comment]
Two more commits. 322392a fixes the issues identified in your comment, but does not respect "job" scoping. 2ad1f98 adds in job scoping, but does it in a somewhat ham-fisted way. I don't expect that it will get merged in, but that it's an early approach to the problem of tracking distinct jobs. I'd like to talk to you about what "jobs" are for Tortoise and how we can make every work well.

**Comment by mrerrormessage** [comment]
Revised to respect agents and instruction (the particular "every"). Job ID is not respected, as that is not a concept in Tortoise (at least until further notice).

**Comment by TheBizzle** [comment]
If you [check out the build](https://travis-ci.org/NetLogo/Tortoise/builds/49771466), `depend` is actually not running correctly.

**Comment by mrerrormessage** [comment]
The output of depend looks the same as [the last build on master](https://travis-ci.org/NetLogo/Tortoise/builds/44488649) to me. Am I missing something?

**Comment by TheBizzle** [comment]
My bad.  Looks like the exception-throwing nonsense is expected, since Classycle attempts to terminate the JVM its running in after checking the dependencies.

**Comment by TheBizzle** [comment]
@mrerrormessage, after some thorough rebasing, this will be ready for a merge.

**Comment by mrerrormessage** [comment]
Rebased down into bite-sized commits. Note that this is getting merged into wip-mitosis, if we want to merge to master that will need to be done separately.

**Comment by TheBizzle** [comment]
Umm... this blew away a bunch of the changes that I made to clean up this branch.

**Comment by TheBizzle** [comment]
I hate to keep dragging this out, but please see the [`wip-bizzle-cleanups`](https://github.com/NetLogo/Tortoise/tree/wip-bizzle-cleanups) branch for a branch with my cleanup commits in it, and incorporate them into these commits.

Is there any reason that we shouldn't just merge `wip-mitosis` once this is ready?

**Comment by mrerrormessage** [comment]
Your changes have been added into the branch.

**Comment by TheBizzle** [comment]
The idea was that they should be rebased in.  Clearly, I can cherry-pick commits, if that's all I want.

**Comment by mrerrormessage** [comment]
Rebased in. I don't see any reason why we shouldn't merge wip-mitosis into master once this is merged.
